Add the ability to specify an address to ouro

MajorArkwolf opened the pull request:
## Problem
When using ouro with transmission, the connection to transmission was failing due to the localhost not being able to access transmission. There was no mechanism that I knew of that could resolve this (It's possible I messed up my configuration on nix which resulted in this issue).

## Solution
- Add a new arg called ``address`` when running as transmission (optional)
- If arg ``address`` is supplied with transmission, provide the argument to transmission-remote
- Add the ADDRESS argument to the flake.nix file to allow it to be passed to ouro
- Add a print message to alert the user what address will be used, if one is supplied.

## Testing
- [x] Added ADDRESS with the expected ip to my nix flake and confirmed transmission successfully updated the value.

## Note
There appears to be a bug that when updating the port, if the call to transmission fails, the new port will get set to the current port, even though transmission never actually applied it. Resulting in ouro desyncing the actual port set with transmission.

MajorArkwolf committed:
feat(transmission)!: add support for `address`

- Some setups require setting the listening address manually

MajorArkwolf committed:
chore: bump version

- Because of the new `--address` option for `transmission`

MajorArkwolf committed:
chore(nix): typo

74k1 commented:
Hello @MajorArkwolf 
Firstly, thanks for the PR! :)
I have a few questions though, just so I can understand the initial problem a little better.

> When using ouro with transmission, the connection to transmission was failing due to the localhost not being able to access transmission.

~~I'm not sure I understand what you mean here. Ouro is supposed to run inside the VPN Namespace (where transmission lives) and then supplies transmission with the port.
This project heavily depends on https://github.com/Maroka-chan/VPN-Confinement (if using the nixos module)~~

~~How did you set it up?~~

edit: looking at your changes, I see what you mean now.

> There appears to be a bug that if updating the address to transmission fails, the new port will get set to the current port, even though transmission never applied it. Resulting in ouro desyncing with transmission.

Interesting observation. I'll have to see if I can repro this.

74k1 commented:
In the meantime, while I test this, would you be so kind to update your commits to match https://www.conventionalcommits.org/en/v1.0.0/ and move the changes to a different branch on your fork?

MajorArkwolf commented:
Sorry my reporting of that bug was worded horribly. Might want to reread it as it makes more sense.

Can do regarding the actions requested!

74k1 commented:
My tests seem to have worked. (Updated my transmission port and added ADDRESS to the ouro configuration)

I unfortunately wasn't able to test the desync behaviour though. Good to know and I'll keep an eye on it.

If you have a fix for that as well, we're happy to review another PR!

For simplicity's sake, changing your commits to conventional commits will be enough, to avoid the need of creating a new PR for the branch change.

Thanks again for the contribution!

reo101 commented on src/service.rs:
```diff
@@ -222,6 +223,10 @@ impl Service for Transmission {
     async fn update(&self, ports: &Ports) -> eyre::Result<()> {
         let mut cmd = Command::new("transmission-remote");
+        if let Some(addr) = &self.address {
+            cmd.args([addr]);
```
```suggestion
            cmd.arg(addr);
```

reo101 commented on flake.nix:
```diff
@@ -80,6 +80,14 @@
+                                  type = types.str;
+                                  default = "127.0.0.1:9091";
+                                  description = ''
+                                    The address and port ouro will provide transmission remote for
+                                    it to connect to. Can be <IP>:<Port> or just <IP>.
```
```suggestion
                                    The address and port that ouro will provide to `transmission-remote`
                                    for it to connect to. Can be <IP>:<Port> or just <IP>.
```

reo101 commented on flake.nix:
```diff
@@ -80,6 +80,14 @@
                                   type = types.str;
                                 };
+                                ADDRESS = lib.mkOption {
+                                  type = types.str;
```
We might want to do some more thorough validation on the `Can be <IP>:<Port> or just <IP>` part here.

reo101 commented on src/service.rs:
```diff
@@ -202,6 +202,7 @@ impl Service for Slskd {
 #[derive(Debug, Clone)]
 pub struct Transmission {
     pub rpc_credentials: Option<Credentials>,
+    pub address: Option<String>,
```
Maybe same here, an `IP` (with an optional port) `struct`/`enum` would be best, but it's not the end of the world if it's just a `String` (the `transmission-remote` call will yell at us then.

reo101 commented on src/main.rs:
```diff
@@ -62,8 +67,14 @@ async fn main() -> eyre::Result<()> {
+
             ServiceData::Transmission(Transmission {
                 rpc_credentials,
+                address: address.clone(),
```
We might stop pattern matching on references here, since we do not need the `args.service` after the `match` itself, thus eliminating the need for this `.clone()`.

reo101 reviewed (commented):
Great job, just a few comments that I think are worth discussing.

reo101 commented:
If you're comfortable, I could apply the proposed changes and do the renaming/reorganization of the commits here.

MajorArkwolf commented:
I don’t mind, go for it

reo101 committed:
chore(nix): reword `ADDRESS` option description

reo101 committed:
chore: prefer `.arg` to `.args` with a singleton

reo101 committed:
feat(cli): move global arguments to a `struct`

- Consists of `gateway` and `vpn_interface`
  (are `#[clap(flatten)]`-ed back into the `Cli` `struct`)
- Allows for less `Clone::clone`-ing

74k1 reviewed (approved):
still works for me :+1:

MajorArkwolf commented:
Cheers for adding this :)
